File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/FaceDetector.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebasemlvision;

import androidx.annotation.NonNull;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.common.FirebaseVisionImage;
import com.google.firebase.ml.vision.common.FirebaseVisionPoint;
import com.google.firebase.ml.vision.face.FirebaseVisionFace;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceContour;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceDetector;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceDetectorOptions;
import com.google.firebase.ml.vision.face.FirebaseVisionFaceLandmark;
import io.flutter.plugin.common.MethodChannel;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class FaceDetector implements Detector {
  private final FirebaseVisionFaceDetector detector;

  FaceDetector(FirebaseVision vision, Map<String, Object> options) {
    detector = vision.getVisionFaceDetector(parseOptions(options));
  }

  @Override
  public void handleDetection(final FirebaseVisionImage image, final MethodChannel.Result result) {
    detector
        .detectInImage(image)
        .addOnSuccessListener(
            new OnSuccessListener<List<FirebaseVisionFace>>() {
              @Override
              public void onSuccess(List<FirebaseVisionFace> firebaseVisionFaces) {
                List<Map<String, Object>> faces = new ArrayList<>(firebaseVisionFaces.size());
                for (FirebaseVisionFace face : firebaseVisionFaces) {
                  Map<String, Object> faceData = new HashMap<>();

                  faceData.put("left", (double) face.getBoundingBox().left);
                  faceData.put("top", (double) face.getBoundingBox().top);
                  faceData.put("width", (double) face.getBoundingBox().width());
                  faceData.put("height", (double) face.getBoundingBox().height());

                  faceData.put("headEulerAngleY", face.getHeadEulerAngleY());
                  faceData.put("headEulerAngleZ", face.getHeadEulerAngleZ());

                  if (face.getSmilingProbability() != FirebaseVisionFace.UNCOMPUTED_PROBABILITY) {
                    faceData.put("smilingProbability", face.getSmilingProbability());
                  }

                  if (face.getLeftEyeOpenProbability()
                      != FirebaseVisionFace.UNCOMPUTED_PROBABILITY) {
                    faceData.put("leftEyeOpenProbability", face.getLeftEyeOpenProbability());
                  }

                  if (face.getRightEyeOpenProbability()
                      != FirebaseVisionFace.UNCOMPUTED_PROBABILITY) {
                    faceData.put("rightEyeOpenProbability", face.getRightEyeOpenProbability());
                  }

                  if (face.getTrackingId() != FirebaseVisionFace.INVALID_ID) {
                    faceData.put("trackingId", face.getTrackingId());
                  }

                  faceData.put("landmarks", getLandmarkData(face));

                  faceData.put("contours", getContourData(face));

                  faces.add(faceData);
                }

                result.success(faces);
              }
            })
        .addOnFailureListener(
            new OnFailureListener() {
              @Override
              public void onFailure(@NonNull Exception exception) {
                result.error("faceDetectorError", exception.getLocalizedMessage(), null);
              }
            });
  }

  private Map<String, double[]> getLandmarkData(FirebaseVisionFace face) {
    Map<String, double[]> landmarks = new HashMap<>();

    landmarks.put("bottomMouth", landmarkPosition(face, FirebaseVisionFaceLandmark.MOUTH_BOTTOM));
    landmarks.put("leftCheek", landmarkPosition(face, FirebaseVisionFaceLandmark.LEFT_CHEEK));
    landmarks.put("leftEar", landmarkPosition(face, FirebaseVisionFaceLandmark.LEFT_EAR));
    landmarks.put("leftEye", landmarkPosition(face, FirebaseVisionFaceLandmark.LEFT_EYE));
    landmarks.put("leftMouth", landmarkPosition(face, FirebaseVisionFaceLandmark.MOUTH_LEFT));
    landmarks.put("noseBase", landmarkPosition(face, FirebaseVisionFaceLandmark.NOSE_BASE));
    landmarks.put("rightCheek", landmarkPosition(face, FirebaseVisionFaceLandmark.RIGHT_CHEEK));
    landmarks.put("rightEar", landmarkPosition(face, FirebaseVisionFaceLandmark.RIGHT_EAR));
    landmarks.put("rightEye", landmarkPosition(face, FirebaseVisionFaceLandmark.RIGHT_EYE));
    landmarks.put("rightMouth", landmarkPosition(face, FirebaseVisionFaceLandmark.MOUTH_RIGHT));

    return landmarks;
  }

  private Map<String, List<double[]>> getContourData(FirebaseVisionFace face) {
    Map<String, List<double[]>> contours = new HashMap<>();

    contours.put("allPoints", contourPosition(face, FirebaseVisionFaceContour.ALL_POINTS));
    contours.put("face", contourPosition(face, FirebaseVisionFaceContour.FACE));
    contours.put("leftEye", contourPosition(face, FirebaseVisionFaceContour.LEFT_EYE));
    contours.put(
        "leftEyebrowBottom", contourPosition(face, FirebaseVisionFaceContour.LEFT_EYEBROW_BOTTOM));
    contours.put(
        "leftEyebrowTop", contourPosition(face, FirebaseVisionFaceContour.LEFT_EYEBROW_TOP));
    contours.put(
        "lowerLipBottom", contourPosition(face, FirebaseVisionFaceContour.LOWER_LIP_BOTTOM));
    contours.put("lowerLipTop", contourPosition(face, FirebaseVisionFaceContour.LOWER_LIP_TOP));
    contours.put("noseBottom", contourPosition(face, FirebaseVisionFaceContour.NOSE_BOTTOM));
    contours.put("noseBridge", contourPosition(face, FirebaseVisionFaceContour.NOSE_BRIDGE));
    contours.put("rightEye", contourPosition(face, FirebaseVisionFaceContour.RIGHT_EYE));
    contours.put(
        "rightEyebrowBottom",
        contourPosition(face, FirebaseVisionFaceContour.RIGHT_EYEBROW_BOTTOM));
    contours.put(
        "rightEyebrowTop", contourPosition(face, FirebaseVisionFaceContour.RIGHT_EYEBROW_TOP));
    contours.put(
        "upperLipBottom", contourPosition(face, FirebaseVisionFaceContour.UPPER_LIP_BOTTOM));
    contours.put("upperLipTop", contourPosition(face, FirebaseVisionFaceContour.UPPER_LIP_TOP));

    return contours;
  }

  private double[] landmarkPosition(FirebaseVisionFace face, int landmarkInt) {
    FirebaseVisionFaceLandmark landmark = face.getLandmark(landmarkInt);
    if (landmark != null) {
      return new double[] {landmark.getPosition().getX(), landmark.getPosition().getY()};
    }

    return null;
  }

  private List<double[]> contourPosition(FirebaseVisionFace face, int contourInt) {
    FirebaseVisionFaceContour contour = face.getContour(contourInt);
    if (contour != null) {
      List<FirebaseVisionPoint> contourPoints = contour.getPoints();
      List<double[]> result = new ArrayList<double[]>();

      for (int i = 0; i < contourPoints.size(); i++) {
        result.add(new double[] {contourPoints.get(i).getX(), contourPoints.get(i).getY()});
      }

      return result;
    }

    return null;
  }

  private FirebaseVisionFaceDetectorOptions parseOptions(Map<String, Object> options) {
    int classification =
        (boolean) options.get("enableClassification")
            ? FirebaseVisionFaceDetectorOptions.ALL_CLASSIFICATIONS
            : FirebaseVisionFaceDetectorOptions.NO_CLASSIFICATIONS;

    int landmark =
        (boolean) options.get("enableLandmarks")
            ? FirebaseVisionFaceDetectorOptions.ALL_LANDMARKS
            : FirebaseVisionFaceDetectorOptions.NO_LANDMARKS;

    int contours =
        (boolean) options.get("enableContours")
            ? FirebaseVisionFaceDetectorOptions.ALL_CONTOURS
            : FirebaseVisionFaceDetectorOptions.NO_CONTOURS;

    int mode;
    switch ((String) options.get("mode")) {
      case "accurate":
        mode = FirebaseVisionFaceDetectorOptions.ACCURATE;
        break;
      case "fast":
        mode = FirebaseVisionFaceDetectorOptions.FAST;
        break;
      default:
        throw new IllegalArgumentException("Not a mode:" + options.get("mode"));
    }

    FirebaseVisionFaceDetectorOptions.Builder builder =
        new FirebaseVisionFaceDetectorOptions.Builder()
            .setClassificationMode(classification)
            .setLandmarkMode(landmark)
            .setContourMode(contours)
            .setMinFaceSize((float) ((double) options.get("minFaceSize")))
            .setPerformanceMode(mode);

    if ((boolean) options.get("enableTracking")) {
      builder.enableTracking();
    }

    return builder.build();
  }

  @Override
  public void close() throws IOException {
    detector.close();
  }
}

File: packages/firebase_ml_vision/ios/Classes/FaceDetector.m
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FirebaseMlVisionPlugin.h"

@interface FaceDetector ()
@property FIRVisionFaceDetector *detector;
@end

@implementation FaceDetector
- (instancetype)initWithVision:(FIRVision *)vision options:(NSDictionary *)options {
  self = [super init];
  if (self) {
    _detector = [vision faceDetectorWithOptions:[FaceDetector parseOptions:options]];
  }
  return self;
}

- (void)handleDetection:(FIRVisionImage *)image result:(FlutterResult)result {
  [_detector
      processImage:image
        completion:^(NSArray<FIRVisionFace *> *_Nullable faces, NSError *_Nullable error) {
          if (error) {
            [FLTFirebaseMlVisionPlugin handleError:error result:result];
            return;
          } else if (!faces) {
            result(@[]);
            return;
          }

          NSMutableArray *faceData = [NSMutableArray array];
          for (FIRVisionFace *face in faces) {
            id smileProb = face.hasSmilingProbability ? @(face.smilingProbability) : [NSNull null];
            id leftProb =
                face.hasLeftEyeOpenProbability ? @(face.leftEyeOpenProbability) : [NSNull null];
            id rightProb =
                face.hasRightEyeOpenProbability ? @(face.rightEyeOpenProbability) : [NSNull null];

            NSDictionary *data = @{
              @"left" : @(face.frame.origin.x),
              @"top" : @(face.frame.origin.y),
              @"width" : @(face.frame.size.width),
              @"height" : @(face.frame.size.height),
              @"headEulerAngleY" : face.hasHeadEulerAngleY ? @(face.headEulerAngleY)
                                                           : [NSNull null],
              @"headEulerAngleZ" : face.hasHeadEulerAngleZ ? @(face.headEulerAngleZ)
                                                           : [NSNull null],
              @"smilingProbability" : smileProb,
              @"leftEyeOpenProbability" : leftProb,
              @"rightEyeOpenProbability" : rightProb,
              @"trackingId" : face.hasTrackingID ? @(face.trackingID) : [NSNull null],
              @"landmarks" : @{
                @"bottomMouth" : [FaceDetector getLandmarkPosition:face
                                                          landmark:FIRFaceLandmarkTypeMouthBottom],
                @"leftCheek" : [FaceDetector getLandmarkPosition:face
                                                        landmark:FIRFaceLandmarkTypeLeftCheek],
                @"leftEar" : [FaceDetector getLandmarkPosition:face
                                                      landmark:FIRFaceLandmarkTypeLeftEar],
                @"leftEye" : [FaceDetector getLandmarkPosition:face
                                                      landmark:FIRFaceLandmarkTypeLeftEye],
                @"leftMouth" : [FaceDetector getLandmarkPosition:face
                                                        landmark:FIRFaceLandmarkTypeMouthLeft],
                @"noseBase" : [FaceDetector getLandmarkPosition:face
                                                       landmark:FIRFaceLandmarkTypeNoseBase],
                @"rightCheek" : [FaceDetector getLandmarkPosition:face
                                                         landmark:FIRFaceLandmarkTypeRightCheek],
                @"rightEar" : [FaceDetector getLandmarkPosition:face
                                                       landmark:FIRFaceLandmarkTypeRightEar],
                @"rightEye" : [FaceDetector getLandmarkPosition:face
                                                       landmark:FIRFaceLandmarkTypeRightEye],
                @"rightMouth" : [FaceDetector getLandmarkPosition:face
                                                         landmark:FIRFaceLandmarkTypeMouthRight],
              },
              @"contours" : @{
                @"allPoints" : [FaceDetector getContourPoints:face contour:FIRFaceContourTypeAll],
                @"face" : [FaceDetector getContourPoints:face contour:FIRFaceContourTypeFace],
                @"leftEye" : [FaceDetector getContourPoints:face contour:FIRFaceContourTypeLeftEye],
                @"leftEyebrowBottom" :
                    [FaceDetector getContourPoints:face
                                           contour:FIRFaceContourTypeLeftEyebrowBottom],
                @"leftEyebrowTop" :
                    [FaceDetector getContourPoints:face contour:FIRFaceContourTypeLeftEyebrowTop],
                @"lowerLipBottom" :
                    [FaceDetector getContourPoints:face contour:FIRFaceContourTypeLowerLipBottom],
                @"lowerLipTop" : [FaceDetector getContourPoints:face
                                                        contour:FIRFaceContourTypeLowerLipTop],
                @"noseBottom" : [FaceDetector getContourPoints:face
                                                       contour:FIRFaceContourTypeNoseBottom],
                @"noseBridge" : [FaceDetector getContourPoints:face
                                                       contour:FIRFaceContourTypeNoseBridge],
                @"rightEye" : [FaceDetector getContourPoints:face
                                                     contour:FIRFaceContourTypeRightEye],
                @"rightEyebrowBottom" :
                    [FaceDetector getContourPoints:face
                                           contour:FIRFaceContourTypeRightEyebrowBottom],
                @"rightEyebrowTop" :
                    [FaceDetector getContourPoints:face contour:FIRFaceContourTypeRightEyebrowTop],
                @"upperLipBottom" :
                    [FaceDetector getContourPoints:face contour:FIRFaceContourTypeUpperLipBottom],
                @"upperLipTop" : [FaceDetector getContourPoints:face
                                                        contour:FIRFaceContourTypeUpperLipTop],
              }
            };

            [faceData addObject:data];
          }

          result(faceData);
        }];
}

+ (id)getLandmarkPosition:(FIRVisionFace *)face landmark:(FIRFaceLandmarkType)landmarkType {
  FIRVisionFaceLandmark *landmark = [face landmarkOfType:landmarkType];
  if (landmark) {
    return @[ landmark.position.x, landmark.position.y ];
  }

  return [NSNull null];
}

+ (id)getContourPoints:(FIRVisionFace *)face contour:(FIRFaceContourType)contourType {
  FIRVisionFaceContour *contour = [face contourOfType:contourType];
  if (contour) {
    NSArray<FIRVisionPoint *> *contourPoints = contour.points;
    NSMutableArray *result = [[NSMutableArray alloc] initWithCapacity:[contourPoints count]];
    for (int i = 0; i < [contourPoints count]; i++) {
      FIRVisionPoint *point = [contourPoints objectAtIndex:i];
      [result insertObject:@[ point.x, point.y ] atIndex:i];
    }
    return [result copy];
  }

  return [NSNull null];
}

+ (FIRVisionFaceDetectorOptions *)parseOptions:(NSDictionary *)optionsData {
  FIRVisionFaceDetectorOptions *options = [[FIRVisionFaceDetectorOptions alloc] init];

  NSNumber *enableClassification = optionsData[@"enableClassification"];
  if (enableClassification.boolValue) {
    options.classificationMode = FIRVisionFaceDetectorClassificationModeAll;
  } else {
    options.classificationMode = FIRVisionFaceDetectorClassificationModeNone;
  }

  NSNumber *enableLandmarks = optionsData[@"enableLandmarks"];
  if (enableLandmarks.boolValue) {
    options.landmarkMode = FIRVisionFaceDetectorLandmarkModeAll;
  } else {
    options.landmarkMode = FIRVisionFaceDetectorLandmarkModeNone;
  }

  NSNumber *enableContours = optionsData[@"enableContours"];
  if (enableContours.boolValue) {
    options.contourMode = FIRVisionFaceDetectorContourModeAll;
  } else {
    options.contourMode = FIRVisionFaceDetectorContourModeNone;
  }

  NSNumber *enableTracking = optionsData[@"enableTracking"];
  options.trackingEnabled = enableTracking.boolValue;

  NSNumber *minFaceSize = optionsData[@"minFaceSize"];
  options.minFaceSize = [minFaceSize doubleValue];

  NSString *mode = optionsData[@"mode"];
  if ([mode isEqualToString:@"accurate"]) {
    options.performanceMode = FIRVisionFaceDetectorPerformanceModeAccurate;
  } else if ([mode isEqualToString:@"fast"]) {
    options.performanceMode = FIRVisionFaceDetectorPerformanceModeFast;
  }

  return options;
}
@end
